# Import Python libraries

In [20]:
import numpy as np
from numpy import nan
import pandas as pd
import re
import requests
import htmlmin

In [2]:
def to_str(item):
    return str(item)

# Get and prepare the data

In [3]:
# Read in license and terms metadata, making sure that all values are string values
licenseMetadataDF = pd.read_csv(
    'licenses_and_terms_metadata_hdv_2022-10-03.csv',
    sep=',', na_filter=False,
    converters={'dataset_version_number': to_str})

# Preview data
licenseMetadataDF.head()

dataset_pid                     dataset_pid_url  \
0  doi:10.7910/DVN/5PRYPC  https://doi.org/10.7910/DVN/5PRYPC   
1   doi:10.7910/DVN/VIE1H   https://doi.org/10.7910/DVN/VIE1H   
2  doi:10.7910/DVN/KKUJWW  https://doi.org/10.7910/DVN/KKUJWW   
3  doi:10.7910/DVN/5E6GBN  https://doi.org/10.7910/DVN/5E6GBN   
4  doi:10.7910/DVN/9MKISZ  https://doi.org/10.7910/DVN/9MKISZ   

  dataset_version_number license_name  \
0                    4.0      CC0 1.0   
1                  248.0                
2                    1.0      CC0 1.0   
3                    1.2      CC0 1.0   
4                   10.0      CC0 1.0   

                                        license_uri terms_of_use  \
0  http://creativecommons.org/publicdomain/zero/1.0                
1                                                                  
2  http://creativecommons.org/publicdomain/zero/1.0                
3  http://creativecommons.org/publicdomain/zero/1.0                
4  http://creativecommons.org/publicdomain/zero/1.0                

  confidentiality_declaration special_permissions restrictions  \
0                                                                
1                                                                
2                                                                
3                                                                
4                                                                

  citation_requirements depositor_requirements conditions disclaimer  \
0                                                                      
1                                                                      
2                                                                      
3                                                                      
4                                                                      

  terms_of_access data_access_place original_archive availability_status  \
0                                                                          
1                                                                          
2                                                                          
3                                                                          
4                                                                          

  contact_for_access size_of_collection study_completion  
0                                                         
1                                                         
2                                                         
3                                                         
4

In [4]:
# Read in basic metadata of each dataset, making sure that dates are recognized as dates and that the version number is treated as a string
basicMetadataDF = pd.read_csv(
    'basic_metadata_hdv_2022-10-03.csv',
    sep=',', na_filter=False,
    parse_dates=['dataset_publication_date', 'dataset_version_create_time'],
    converters={'dataset_version_number': to_str})

# Preview data
basicMetadataDF.head()

dataset_pid                     dataset_pid_url  \
0  doi:10.7910/DVN/5PRYPC  https://doi.org/10.7910/DVN/5PRYPC   
1   doi:10.7910/DVN/VIE1H   https://doi.org/10.7910/DVN/VIE1H   
2  doi:10.7910/DVN/KKUJWW  https://doi.org/10.7910/DVN/KKUJWW   
3  doi:10.7910/DVN/5E6GBN  https://doi.org/10.7910/DVN/5E6GBN   
4  doi:10.7910/DVN/9MKISZ  https://doi.org/10.7910/DVN/9MKISZ   

  dataset_version_number dataset_publication_date dataset_version_create_time  \
0                    4.0               2020-05-27   2020-06-17 23:49:50+00:00   
1                  248.0               2015-02-11   2015-06-18 19:49:12+00:00   
2                    1.0               2015-07-23   2015-07-15 00:57:41+00:00   
3                    1.2               2020-04-05   2020-04-05 21:32:37+00:00   
4                   10.0               2011-06-09   2013-01-29 23:42:10+00:00   

  dataset_version_state          publisher  
0              RELEASED  Harvard Dataverse  
1              RELEASED  Harvard Dataverse  
2              RELEASED  Harvard Dataverse  
3              RELEASED  Harvard Dataverse  
4              RELEASED  Harvard Dataverse

In [5]:
# Get the unique counts of datasets in the licenseMetadataDF and datasetsDF dataframes
print('Number of dataset versions licenseMetadataDF: %s' %(len(licenseMetadataDF)))
print('Number of datasets in licenseMetadataDF: %s' % (len(pd.unique(licenseMetadataDF['dataset_pid']))))

print('\nNumber of dataset versions in basicMetadataDF: %s' %(len(basicMetadataDF)))
print('Number of datasets in basicMetadataDF: %s' % (len(pd.unique(basicMetadataDF['dataset_pid']))))

Number of dataset versions licenseMetadataDF: 130768
Number of datasets in licenseMetadataDF: 80278

Number of dataset versions in basicMetadataDF: 130768
Number of datasets in basicMetadataDF: 80278


In [6]:
# Make sure the two dataframes, allDatasetPIDsDF and basicMetadataDF, contain the datasets
datasetsInlicenseMetadataDFList = set(licenseMetadataDF['dataset_pid'].tolist())
datasetsInbasicMetadataDFList = set(basicMetadataDF['dataset_pid'].tolist())

print(f'Datasets in datasetsInlicenseMetadataDFList: {len(datasetsInlicenseMetadataDFList)}')
print(f'Datasets in basicMetadataDF: {len(datasetsInbasicMetadataDFList)}')

# Create list of dataset PIDs that exist in one list and not the other, if any
difference = datasetsInlicenseMetadataDFList - datasetsInbasicMetadataDFList
difference = list(difference)
print(f'Number of datasets in one table and not the other: {len(difference)}')


Datasets in datasetsInlicenseMetadataDFList: 80278
Datasets in basicMetadataDF: 80278
Number of datasets in one table and not the other: 0


In [7]:
# Merge the two dataframes
licenseAndBasicMetadataDF = pd.merge(
    licenseMetadataDF, basicMetadataDF,
    how='inner',
    on=[
        'dataset_pid',
        'dataset_pid_url',
        'dataset_version_number'])

In [8]:
# Check count of unique datasets in the new mergedDF
print(f'Number of dataset versions in licenseAndBasicMetadataDF: {len(licenseAndBasicMetadataDF)}')
uniqueDatasetCount = len(pd.unique(licenseAndBasicMetadataDF['dataset_pid']))
print(f'Number of datasets in licenseAndBasicMetadataDF: {uniqueDatasetCount}')

Number of dataset versions in licenseAndBasicMetadataDF: 130768
Number of datasets in licenseAndBasicMetadataDF: 80278


In [11]:
# Create dataframe, licenseMetadataLatestVersionDF, that contains the license and terms metadata of only the latest
# version of each dataset, using the latest version release time from the merged basicMetadataDF
latestversion_licenseAndBasicMetadataDF = (licenseAndBasicMetadataDF
                 .iloc[licenseAndBasicMetadataDF.groupby('dataset_pid')['dataset_version_create_time']
                 .agg(pd.Series.idxmax)]
                 .reset_index(drop=True, inplace=False))

# Remove dataset_version_state and publisher columns and reorder remaining columns
latestversion_licenseAndBasicMetadataDF = latestversion_licenseAndBasicMetadataDF[[
    'dataset_pid_url', 'dataset_publication_date', 'dataset_version_number', 'dataset_version_create_time', 'license_name', 'license_uri', 'terms_of_use', 'confidentiality_declaration', 'special_permissions', 'restrictions', 'citation_requirements', 'depositor_requirements', 'conditions', 'disclaimer', 'terms_of_access', 'data_access_place', 'original_archive', 'availability_status', 'contact_for_access', 'size_of_collection', 'study_completion'
]]

print(f'Count of pids in latestversion_mergedDF: {len(latestversion_licenseAndBasicMetadataDF)}')


Count of pids in latestversion_mergedDF: 80278


In [12]:
# Let's make sure each column is the object type we want. Everything should be a string except for the two date columns.
latestversion_licenseAndBasicMetadataDF.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 80278 entries, 0 to 80277
Data columns (total 21 columns):
 #   Column                       Non-Null Count  Dtype              
---  ------                       --------------  -----              
 0   dataset_pid_url              80278 non-null  object             
 1   dataset_publication_date     80278 non-null  datetime64[ns]     
 2   dataset_version_number       80278 non-null  object             
 3   dataset_version_create_time  80278 non-null  datetime64[ns, UTC]
 4   license_name                 80278 non-null  object             
 5   license_uri                  80278 non-null  object             
 6   terms_of_use                 80278 non-null  object             
 7   confidentiality_declaration  80278 non-null  object             
 8   special_permissions          80278 non-null  object             
 9   restrictions                 80278 non-null  object             
 10  citation_requirements        80278 non-null  o

# Use the data

Now that we have the data we need, let's answer our question:
- Which standard licences have depositors entered into the Terms fields?

As part of the Dataverse software's v5.10 multiple license update:
1. If a dataset has a CC0 waiver and nothing entered in any of its "Terms of Use" fields, the dataset is considered to have a "standard" CC0 license, and the metadata of those datasets, the "license_name" field has "CC0".
2. If a dataset has a CC0 waiver and anything entered in any of its "Terms of Use" fields, the dataset is considered to have a custom license, where "This dataset is made available under a Creative Commons CC0 license with the following additional/modified terms and conditions:" is entered in the Terms of Use field, followed by whatever text was entered in any of the "Terms of Use" fields.
3. If the dataset does not have a CC0 waiver and anything entered in any of its "Terms of Use" fields, the dataset is considered to have a custom license, where what's   in any of the "Terms of Use" fields makes us the custom license.
4. If the CC0 waiver is not applied to the dataset and nothing is entered in any of the Terms of Use fields, the text "This dataset is made available without information on how it can be used. You should communicate with the Contact(s) specified before use." was entered in the Terms of Use field

To find datasets with "custom licenses", we want to find datasets that match the second or third conditions.

In [17]:
datasetsWithCustomLicense = (latestversion_licenseAndBasicMetadataDF
                     .query('license_name == "" and terms_of_use != "This dataset is made available without information on how it can be used. You should communicate with the Contact(s) specified before use."')
                     .reset_index(drop=True, inplace=False)
                     )

datasetsWithCustomLicense.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32645 entries, 0 to 32644
Data columns (total 21 columns):
 #   Column                       Non-Null Count  Dtype              
---  ------                       --------------  -----              
 0   dataset_pid_url              32645 non-null  object             
 1   dataset_publication_date     32645 non-null  datetime64[ns]     
 2   dataset_version_number       32645 non-null  object             
 3   dataset_version_create_time  32645 non-null  datetime64[ns, UTC]
 4   license_name                 32645 non-null  object             
 5   license_uri                  32645 non-null  object             
 6   terms_of_use                 32645 non-null  object             
 7   confidentiality_declaration  32645 non-null  object             
 8   special_permissions          32645 non-null  object             
 9   restrictions                 32645 non-null  object             
 10  citation_requirements        32645 non-null  o

In [25]:
# For each of these datasets, in order to determine which standard licenses might be mentioned in any of their Terms of Use or Terms of Access fields, let's create a new column that's a concatenation of what's entered in all of the Terms fields. Then we can search using only that that column
termsColumns = [
    'terms_of_use',
    'confidentiality_declaration',
    'special_permissions',
    'restrictions',
    'citation_requirements',
    'depositor_requirements',
    'conditions',
    'disclaimer',
    'terms_of_access',
    'data_access_place',
    'original_archive',
    'availability_status',
    'contact_for_access',
    'size_of_collection',
    'study_completion'
]
datasetsWithCustomLicense['combined_terms'] = datasetsWithCustomLicense[termsColumns].apply(lambda row: ' '.join(row.values.astype(str)), axis=1)

datasetsWithCustomLicense.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32645 entries, 0 to 32644
Data columns (total 22 columns):
 #   Column                       Non-Null Count  Dtype              
---  ------                       --------------  -----              
 0   dataset_pid_url              32645 non-null  object             
 1   dataset_publication_date     32645 non-null  datetime64[ns]     
 2   dataset_version_number       32645 non-null  object             
 3   dataset_version_create_time  32645 non-null  datetime64[ns, UTC]
 4   license_name                 32645 non-null  object             
 5   license_uri                  32645 non-null  object             
 6   terms_of_use                 32645 non-null  object             
 7   confidentiality_declaration  32645 non-null  object             
 8   special_permissions          32645 non-null  object             
 9   restrictions                 32645 non-null  object             
 10  citation_requirements        32645 non-null  o

In [54]:
# Let's get a unique list of what's been entered in the Terms fields
customMetadataList = list(set(datasetsWithCustomLicense['combined_terms'].tolist()))
print(f'Number of unique strings entered in Terms fields: {len(customMetadataList)}')

Number of unique strings entered in Terms fields: 2638


In [55]:
# Let's figure out which custom licenses are in these strings

# Some of the things entered in the Terms fields are HTML. Let's remove the formatting so that it's faster to search through.
for place, metadata in enumerate(customMetadataList):
    metadata = htmlmin.minify(metadata, remove_empty_space=True)
    customMetadataList[place] = metadata

In [83]:
# There's probably a lot of variety in the ways that standard licenses are identified by name, but there might be more standardization in how they're identified by their URLs, so let's extract the URLs from this custom metadata
urlsList = []
regex = r"(?i)\b((?:https?://|www\d{0,3}[.]|[a-z0-9.\-]+[.][a-z]{2,4}/)(?:[^\s()<>]+|\(([^\s()<>]+|(\([^\s()<>]+\)))*\))+(?:\(([^\s()<>]+|(\([^\s()<>]+\)))*\)|[^\s`!()\[\]{};:'\".,<>?«»“”‘’]))"

for metadata in customMetadataList:
    urls = re.search(regex, metadata)
    if urls is not None:
        # Try normalizing the URLs a bit more to get a better count of the unique URLs
        urls = urls.group().replace('http://', 'https://').replace('%2F', '/').replace('%3A', ':')
        urls = urls.rstrip('/')
        urlsList.append(urls)

# Deduplicate the list of URLs
urlsList = list(set(urlsList))

# Print first five items in the list
print(f'First five URLs in the list of {len(urlsList)} URLs:')
for url in urlsList[:5]:
    print(url)

First five urls in list of 225 URLs:
https://www.canvaslms.com/policies/api-policy
https://dataverse.harvard.edu/file.xhtml?fileId=3102931&version=RELEASED&version=.1
https://adsabs.harvard.edu/abs/2008ApJS..175..277D
https://www.socio.com/teenpregnancydata.php
https://dx.doi.org/10.7910/DVN/YZBRYX


In [65]:
# Looks like it worked. Now we just need to find the URLs that point to licenses. This is should be easy for Creative Commons licenses:
creativeCommonsUrls = []
for url in urlsList:
    if 'creativecommons.org/licenses' in url:
        creativeCommonsUrls.append(url)
creativeCommonsUrls.sort()

print(f'Count of Creative Commons URLs: {len(creativeCommonsUrls)}')
for url in creativeCommonsUrls:
    print(url)

Count of Creative Commons URLs: 23
https://creativecommons.org/licenses/by-nc-nd/4.0
https://creativecommons.org/licenses/by-nc-nd/4.0/legalcode
https://creativecommons.org/licenses/by-nc-sa/3.0
https://creativecommons.org/licenses/by-nc-sa/3.0/deed.en
https://creativecommons.org/licenses/by-nc-sa/3.0/us
https://creativecommons.org/licenses/by-nc-sa/4.0
https://creativecommons.org/licenses/by-nc/2.0
https://creativecommons.org/licenses/by-nc/3.0
https://creativecommons.org/licenses/by-nc/4.0
https://creativecommons.org/licenses/by-nc/4.0/legalcode
https://creativecommons.org/licenses/by-nd/4.0
https://creativecommons.org/licenses/by-sa/2.0
https://creativecommons.org/licenses/by-sa/4.0
https://creativecommons.org/licenses/by-sa/4.0/legalcode
https://creativecommons.org/licenses/by/2.0
https://creativecommons.org/licenses/by/2.5/it
https://creativecommons.org/licenses/by/3.0
https://creativecommons.org/licenses/by/3.0/us
https://creativecommons.org/licenses/by/4.0
https://creativecommon

In [66]:
# Looks like there are 23 unique URLs, but there might be some duplicates.
# Some URLs end in "/legalcode", which are just the official text of the license. So let's remove that.
# Some URLs end in "/deed" or "/deed.en". When those parts are removed from those URLs, they point to the same pages, so let's remove those, too.
# And one URL seems to have been generated by Outlook and contains the URL for the BY-NC-SA 3.0 license. That license is already in the list, so let's remove that "safelinks URL.

for place, url in enumerate(creativeCommonsUrls):
    if '/legalcode' or '/deed' in url:
        url = url.replace('/legalcode', '').replace('/deed.en', '').replace('/deed', '')
        creativeCommonsUrls[place] = url
    if 'safelinks' in url:
        creativeCommonsUrls.pop(place)

# Deduplicate the list
creativeCommonsUrls = list(set(creativeCommonsUrls))

print(len(creativeCommonsUrls))
for url in creativeCommonsUrls:
    print(url)

15
https://creativecommons.org/licenses/by-sa/4.0
https://creativecommons.org/licenses/by-sa/2.0
https://creativecommons.org/licenses/by/4.0
https://creativecommons.org/licenses/by-nc-sa/3.0
https://creativecommons.org/licenses/by-nc/3.0
https://creativecommons.org/licenses/by/2.5/it
https://creativecommons.org/licenses/by/3.0/us
https://creativecommons.org/licenses/by-nc/2.0
https://creativecommons.org/licenses/by-nd/4.0
https://creativecommons.org/licenses/by/2.0
https://creativecommons.org/licenses/by-nc-nd/4.0
https://creativecommons.org/licenses/by-nc-sa/4.0
https://creativecommons.org/licenses/by-nc-sa/3.0/us
https://creativecommons.org/licenses/by-nc/4.0
https://creativecommons.org/licenses/by/3.0


So there are 15 different Creative Commons licenses used in datasets published in the Harvard Dataverse Repository.

What about URLs that point to different types of licenses? For that, lets see what URLs are in SPDX, which maintains structured information about standard licenses.

In [67]:
spdxDataUrl = 'https://raw.githubusercontent.com/spdx/license-list-data/master/json/licenses.json'
response = requests.get(spdxDataUrl)
spdxData = response.json()

# Let's save the license information from this list, removing the Creative Commons licenses
licenseDict = []
for license in spdxData['licenses']:
    if 'Creative Commons' not in license['name']:
        licenseDict.append(license)

print(f'Count of licenses in SPDX (excluding Creative Commons licenses:{len(licenseDict)}')

447


## What are the types of licenses used in other Dataverse repositories

447 is a lot of licenses. Trying to find if any of these licenses have been applied to datasets in the Harvard repository might not be worth the work.

Instead, let's see what types of licenses other Dataverse repositories are using. I've collected this information using the Dataverse API and sorted the licenses by type.

Here's a list of the types of licenses used by other Dataverse repositories and the base URLs used for each license.

- Apache: https://www.apache.org/licenses
- BSD: https://opensource.org/licenses
- GNU: https://www.gnu.org/licenses
- MIT: https://opensource.org/licenses/MIT
- Mozilla: https://mozilla.org/MPL

All of these licenses are meant to be applied to software. The Harvard Dataverse Repository requires that depositors always include data, and the recommended way to apply different licenses for different parts of a dataset, such as applying one license for the data files and one for software files, is to create a custom license.
So we should exclude licenses meant only for software.


## What about licenses generally recommended for data?

The Digital Curation Centre maintains a page at https://www.dcc.ac.uk/guidance/how-guides/license-research-data that describes two other types of data licenses.

- Open Data Commons: https://opendatacommons.org/licenses
- Open Government Licence (OGL): https://www.nationalarchives.gov.uk/doc/open-government-licence

Let's see if any of these licenses' URLs are in our list.

In [79]:
openDataCommonsUrls = []
for url in urlsList:
    if 'opendatacommons.org/licenses' in url:
        openDataCommonsUrls.append(url.strip())
openDataCommonsUrls = list(set(openDataCommonsUrls))
openDataCommonsUrls.sort()
print(f'Count of Open Data Commons License URLs: {len(openDataCommonsUrls)}')
for url in openDataCommonsUrls:
    print(url)

openGovernmentLicenseUrls = []
for url in urlsList:
    if 'open-government-licence' in url:
        openGovernmentLicenseUrls.append(url.strip())
openGovernmentLicenseUrls = list(set(openGovernmentLicenseUrls))
openGovernmentLicenseUrls.sort()
print(f'\nCount of Open Government Licence URLs: {len(openGovernmentLicenseUrls)}')
for url in openGovernmentLicenseUrls:
    print(url)

Count of Open Data Commons License URLs: 4
https://opendatacommons.org/licenses/by/1.0
https://opendatacommons.org/licenses/odbl
https://opendatacommons.org/licenses/odbl/1-0
https://opendatacommons.org/licenses/odbl/1.0

Count of Open Government Licence URLs: 1
https://open.canada.ca/en/open-government-licence-canada


Two Open Data Commons licenses are used. Since there are only three, we should consider adding all three to the list: https://opendatacommons.org/licenses

It looks like there are no Open Government Licences from the UK's National Archives, but by searching for URLs with "open-government-license", we found that Canada's   Open Government License has been used.

# Questions and recommendations

## Questions
- I think we should include Creative Commons licenses, but should we include their older versions? I think so, if only so that the datasets with those license can be updated with new versions in order to pick the licenses from the list, making the licenses more machine readable.
- For newer datasets, is it recommended to use older versions of Creative Commons licenses? If not, and if those older licenses are among the options that depositors can choose from, how can we discourage the use of the older versions or encourage the use of the latest versions?

## Recommendations
- We should include the three Open Data Commons Licenses (https://opendatacommons.org/licenses)
- We should include Canada's Open Government Licences